### ISSUE TITLE
Non VIP Atrium doors can be inverted

### ISSUE THREAD

**Issue by FranchNotFrench** [paulkim26/collab-museum#221]
Zesty mentioned this in the VIP doors Issue but that is closed now, so splitting into a separate issue.

While difficult to reproduce, if player(s) quickly trigger both the On Enter and On Exit actions tied to door animations, they can invert.

Potential solutions:
- When On Enter and/or On Exit is triggered, briefly disable the trigger (need to make sure it doesn't trigger an action when it enables though)
- Split into two differently sized triggers, one for On Enter and one for On Exit, so one player can't trip both simultaneously. Not sure which would be bigger...whatever feels intuitive.

**Comment by FranchNotFrench** [comment]
Yeah, splitting this into two differently sized triggers tied to separate On Enter / On Exit events was a horrible idea lol. I should have listened to the voice in the back of my head that said "You might want to think through the logic of that before suggesting it." The issue, of course, is that if On Enter is smaller, you can trigger it repeatedly and desync without ever leaving the On Exit trigger. Or vice versa.

I'll test an enable/disable toggle delay. It might work as long as triggers never treat enable/disable as an On Enter / On Exit event.

**Comment by FranchNotFrench** [comment]
This seems to work great and I am going to implement on all six non VIP doors. I expanded the door triggers slightly to compensate for the password check delay of 0.1.

**Comment by FranchNotFrench** [comment]
Somehow I was still able to Invert the latest version of a Non VIP door, and it stayed inverted. I think it was one of the Creator Card doors but I'm guessing they are all the same. I couldn't figure out what happened and couldn't reproduce it. Reopening for now.

The current logic is:
Trigger On Enter: pass value to a lock with password 1, that has 0.1 second delay on checking password
Lock On Unlock (Receives 1 from Trigger): Trigger door animation
Lock On Lock (Receives 0 from Trigger): Trigger door animation

I'm still not seeing how this could invert, given the lock check password delay. It should be impossible for a single player to simultaneously trigger multiple On Enter events. If you trigger On Enter sequentially, even in rapid succession, that means you had to exit between them which should throw a 1-0-1 to the lock but separated by at least 0.1 seconds each).

**Comment by FranchNotFrench** [comment]
I managed to get the doors to invert once today. But for the life of me I can't figure out how to reliably recreate it, and the logic doesn't make any sense to me either. I'm pretty sure this is happening with different doors and that they all have the same properties, rather than one that is incorrect. Hulp.

**Comment by FranchNotFrench** [comment]
I have not yet tried playing with the "check password on delay" lock settings which are still at 0.1 seconds. But without being able to reliably reproduce this bug, it's just random guessing to change the delay and wonder if you made it better or worse.

**Comment by paulkim26** [comment]
The nuclear option would be to make doors open on trigger permanently if their behaviour can't be guaranteed.

**Comment by FranchNotFrench** [comment]
Yuck lol

**Comment by FranchNotFrench** [comment]
The less nuclear option would be for Wollo to modify the VIP door Lua to accommodate the remaining doors. Buy him a Radler or something.

**Comment by FranchNotFrench** [comment]
I gave this some thought, and I think I have a theory of what is happening. Or at least, why the lock + check password delay isn't actually doing anything useful...

The bounced door animation assumes that every time you enter the trigger it will fire once to open, and every time you exit the trigger it will fire a second time to close. When these don't happen in pairs, that is when it inverts (stating the obvious).

I put the lock in with a delay thinking that the issue was someone simultaneously entering and exiting the trigger from the engine's perspective, thereby confusing the animation and inverting it. However, I don't think that is what is happening. Or at least, the lock doesn't do anything to improve it.

With a 0.1 second delay on checking the password (which triggers the door animation on both unlock (1) and lock (0)), suppose you enter the trigger and the lock is sitting on a 1, but then you exit the trigger 0.05 seconds later before the password check occurs, so now the lock is sitting on a 0 again. If nothing else changes in the next 0.1 seconds, the door animation will fire once when the lock locks, causing the door to open but you will be outside the trigger. Which is bad.

Extending the check password delay would probably make things even worse - it would raise the possibility you could send the lock an even number of commands during the delay, but resulting in a single animation = inversion.

If that is the case, then best case scenario would be to eliminate the password check delay, but at that point it seems like the lock isn't doing anything that the trigger couldn't do by itself - just connect On Enter and On Exit to the animation, which is what I originally had, and what wasn't working.

TL;DR - I want to eliminate the locks since they don't solve the original issue. But, I imagine going back to the original On Enter / On Exit door animation triggers will still introduce inversion.

I might take another run at two differently sized triggers - I think this might have not been working before since I wasn't toggling their visibility sequentially. In my head, this could work:

1) Small On Enter trigger that triggers the door animation On Enter then de-activates itself and activates a larger On Exit trigger, ensuring that you will still be in the On Exit trigger when you trigger On Enter.

2) On Exit trigger also triggers the door animation then de-activates itself and activates the smaller On Enter trigger, ensuring that as soon as you exit the On Exit trigger, you are not close enough to the On Enter trigger to simultaneously enter it.

Not sure if this will work, or what it means for multiplayer. But it's the best I have come up with. :)

**Comment by FranchNotFrench** [comment]
I might have this working...using Viz activators to toggle the On Enter and On Exit triggers didn't work, since it doesn't registers as a new On Enter event when the triggers Enables where the player is standing (so you have to first exit and re-enter it to get it to trigger, which doesn't work), but animating the triggers in/out of the player hitbox does appear to work to force an Action each time and to reset the On Enter trigger when the On Exit trigger is active.

I'm think I'm going to replace all the non VIP Atrium doors with this logic.
